src/UI/PlayerSetupInfo.java

```diff
@@ -17,61 +15,74 @@
-      for( int co : coList )
-        coString.append(co).append(" ");
-      coString.append(")");
+      coString.append(deets.co).append(" ");
+      colors.  append(deets.color).append(" ");
+      factions.append(deets.faction).append(" ");
     }
-    else
-      coString.append(coList.get(0));
-    return String.format("%s %s %s %s %s %s", coString, currentColor, currentFaction, flipUnits, currentTeam, currentAi);
+
+    return String.format("%s %s %s %s %s %s %s", coCount, coString, colors, factions, flipUnits, currentTeam, currentAi);
   }
   /** Initializes based on the schema defined by the toString() method above */
   public void initFromString(String input)
   {
-    Scanner s = new Scanner(input);
+    String noParens = input.replace("\\(\\)", "");
+    // Throw away any starting space since it makes us barf
+    if( noParens.charAt(0) == ' ' )
+      noParens = noParens.substring(1);
+    String[] s = noParens.split("\\s+");
+    int si = 0;
 
     coList.clear();
-    // If we found parentheses, then we've got a list to read
-    final Pattern listRegex = Pattern.compile("\\(.*\\)");
-    String coString = s.findInLine(listRegex);
-    if( null != coString )
+    int coCount = 1;
+    if( s.length > si )
+      coCount = Integer.valueOf(s[si++]);
+    // If we wouldn't have added a count or it's too short to have everything, assume there's no prepended CO count
+    if( coCount < 2 || input.length() < 11 + 6 * coCount )
     {
-      Scanner coScanner = new Scanner(coString);
-      coScanner.skip("\\(");
-      while (coScanner.hasNextInt())
-      {
-        coList.add(coScanner.nextInt() % availableCommanders.length);
-      }
-      coScanner.close();
+      coCount = 1;
+      si = 0;
     }
-    else if( s.hasNextInt() )
-      coList.add(s.nextInt() % availableCommanders.length);
-
-    if( s.hasNextInt())
-      currentColor = s.nextInt() % availableColors.length;
-    if( s.hasNextInt())
-      currentFaction = s.nextInt() % availableFactions.length;
-    if( s.hasNextBoolean() )
-      flipUnits = s.nextBoolean();
-    if( s.hasNextInt())
-      currentTeam = s.nextInt();
-    if( s.hasNextInt())
-      currentAi = s.nextInt() % availableAis.length;
-    s.close();
+
+    for(int i = 0; i < coCount; ++i)
+      coList.add(new CODeets());
+
+    for( int i = 0; i < coCount; ++i )
+      if( s.length > si )
+        coList.get(i).co = Integer.valueOf(s[si++]) % availableCommanders.length;
+    for( int i = 0; i < coCount; ++i )
+      if( s.length > si )
+        coList.get(i).color = Integer.valueOf(s[si++]) % availableColors.length;
+    for( int i = 0; i < coCount; ++i )
+      if( s.length > si )
+        coList.get(i).faction = Integer.valueOf(s[si++]) % availableFactions.length;
+
+    if( s.length > si )
+      flipUnits = Boolean.valueOf(s[si++]);
+    if( s.length > si )
+      currentTeam = Integer.valueOf(s[si++]);
+    if( s.length > si )
+      currentAi = Integer.valueOf(s[si++]) % availableAis.length;
```

Comment:
So each member of a tag team can have their own art and color? Having everything change color when you hit `SWAP` is... strange, to say the least. I'd probably prefer to have one color/faction per player for standard tag teams (though I know you have ideas for other tag modes).